python/mujinwebstackclient/webstackclient.py

```diff
@@ -78,6 +78,80 @@ def _FormatHTTPDate(dt):
     month = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'][dt.month - 1]
     return '%s, %02d %s %04d %02d:%02d:%02d GMT' % (weekday, dt.day, month, dt.year, dt.hour, dt.minute, dt.second)
 
+def BreakLargeQuery(queryFunction):
+    """This decorator break a large query into a few small queries to prevent webstack from consuming too much memory.
+    """
+    def inner(self, *args, **kwargs):
+        if kwargs.get('limit', 0) != 0:
+            return queryFunction(self, *args, **kwargs)
+        
+        initialOffset = kwargs.get('offset', 0)
+        iterator = QueryIterator(queryFunction, *((self,) + args), **kwargs)
+        data = [item for item in iterator]
+        return WebstackClient.ObjectsWrapper({'objects': data, 'meta':{'total_count': iterator.totalCount, 'limit':0, 'offset': initialOffset}})
+
+    return inner    
+
+class QueryIterator:
+    """Converts a large query to a iterator. The iterator will internally query webstack with a few small queries
+    example:
+
+      iterator = QueryIterator(client.GetScenes)
+      iterator = QueryIterator(client.GetScenes, offset=10, limit=10)
+      for scene in QueryIterator(client.GetScenes, offset=10, limit=10):
+          do_something(scene)
+    """
+
+    _queryFunction = None
+    _args = None
+    _kwargs = None
+    _items = None
+    _shouldStop = None
+    _totalLimit = None
+    _count = None
+    _totalCount = None
+
+    def __init__(self, queryFunction, *args, **kwargs):
+        self._queryFunction = queryFunction
+        self._args = args
+        self._kwargs = kwargs
+        self._items = []
+        self._shouldStop = False
+        self._kwargs.setdefault('offset', 0)
+        self._kwargs.setdefault('limit', 0)
+        self._totalLimit = self._kwargs['limit']
+        if self._totalLimit == 0:
+            self._totalLimit = 9999999999999 # 0 means no limit
+        self._count = 0
+        self._kwargs['limit'] = 100
```

Comment:
Isn't this overwriting the limit supplied in the constructor?

Reply:
This is intentional. I store what's provided in the constructor in `self._totalLimit`. Here 100 is the limit we set on webstack

But sure, It can be optimized to `self._kwargs['limit'] = min(self._kwargs['limit'], 100)` so that when user only wants 30 items, the client only query webstack for 30 items.